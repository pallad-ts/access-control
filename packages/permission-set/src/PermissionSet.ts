import {EntityRef, Factory} from "@pallad/entity-ref";
import {AccessQuery, Policy} from "@pallad/access-control";

export class PermissionSet<T extends [string, EntityRef<string, {}>]> {
    private entries = new Map<string, Set<Factory<string, any>>>();
    private entityRefTypes = new Set<string>();

    readonly type!: T;

    constructor(readonly subjectType: string) {

    }

    canHandleQuery({action, subject}: AccessQuery) {
        return this.isValidSubject(subject) && this.isValidEntityRefForAction(action, subject.entityRef);
    }

    private isValidEntityRefForAction(action: string, entityRef: EntityRef<string, any>) {
        const entityRefFactoriesForPrivilege = this.entries.get(action);
        return Array.from(entityRefFactoriesForPrivilege ?? []).some(x => x.is(entityRef));
    }

    private isValidSubject(subject: unknown): subject is PermissionSet.Subject<T[1]> {
        // eslint-disable-next-line no-null/no-null
        return typeof subject === 'object' && subject !== null && 'subjectType' in subject && (subject as any).subjectType === this.subjectType;
    }

    wrapEntityRefWithSubject<TEntityRef extends T[1]>(entityRef: TEntityRef): PermissionSet.Subject<TEntityRef> {
        return Object.freeze({
            entityRef,
            subjectType: this.subjectType
        });
    }

    addEntry<
        TAction extends string,
        TEntityFactory extends Factory<string, any>
    >(action: TAction, entityRefFactory: TEntityFactory) {
        this.entityRefTypes.add(entityRefFactory.typeName);
        if (!this.entries.has(action)) {
            this.entries.set(action, new Set());
        }
        this.entries.get(action)!.add(entityRefFactory);
        return this;
    }

    createPolicy<TPrincipal>(
        policyFunction: Policy<PermissionSet.PolicyAccessQueryInput<TPrincipal, T>>
    ): Policy<PermissionSet.PolicyAccessQueryOutput<TPrincipal, T>> {
        return (query: PermissionSet.PolicyAccessQueryOutput<TPrincipal, T>) => {
            if (this.canHandleQuery(query)) {
                return policyFunction({
                    action: query.action,
                    subject: query.subject.entityRef,
                    principal: query.principal
                } as any);
            }
        }
    }
}

export namespace PermissionSet {
    export type PolicyAccessQueryInput<TPrincipal, T extends [string, EntityRef<string, {}>]> =
        T extends [infer TAction, infer TEntityRef] ?
            (TAction extends string ?
                (TEntityRef extends EntityRef<string, {}> ? AccessQuery<TPrincipal, TAction, TEntityRef> : AccessQuery<TPrincipal>) :
                AccessQuery<TPrincipal>
                ) : AccessQuery<TPrincipal>;

    export type PolicyAccessQueryOutput<TPrincipal, T extends [string, EntityRef<string, {}>]> =
        T extends [infer TAction, infer TEntityRef] ?
            (TAction extends string ?
                (TEntityRef extends EntityRef<string, {}> ? AccessQuery<TPrincipal, TAction, Subject<TEntityRef>> : AccessQuery<TPrincipal>) :
                AccessQuery<TPrincipal>
                ) : AccessQuery<TPrincipal>;

    export interface Subject<TEntityRef extends EntityRef<string, {}>> {
        readonly subjectType: string;
        readonly entityRef: TEntityRef;
    }

    export interface AccessQueryParams<TAction extends string, TEntityRef extends EntityRef<string, {}>> {
        action: TAction;
        subject: Subject<TEntityRef>;
    }
}


