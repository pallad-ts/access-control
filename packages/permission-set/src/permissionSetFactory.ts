import {Factory} from "@pallad/entity-ref";
import {PermissionSet} from "./PermissionSet";

export function permissionSetFactory<T extends Record<string, readonly [string, Factory<string, any>]>>(name: string, methods: T) {

    const permissionSet = new PermissionSet(name);

    const result: any = {
        permissionSet,
    };

    for (const [method, [action, entityRefFactory]] of Object.entries(methods)) {
        result[method] = (entityRef: any): PermissionSet.AccessQueryParams<string, any> => {
            return {action, subject: permissionSet.wrapEntityRefWithSubject(entityRef)};
        }
    }

    return result as permissionSetFactory.PermissionSetPreset<T>;
}

export namespace permissionSetFactory {
    export type PermissionSetPreset<T extends Record<string, readonly [string, Factory<string, any>]>> = {
        permissionSet: PermissionSet<MethodsToAccessQueryParamsUnion<T[keyof T]>>,
    } & {
        [TKey in keyof T]: (ref: ReturnType<T[TKey][1]>) => PermissionSet.AccessQueryParams<T[TKey][0], ReturnType<T[TKey][1]>>
    }
=
    export type MethodsToAccessQueryParamsUnion<T extends readonly [string, Factory<string, any>]> =
        T extends readonly [infer TAction, infer TEntityRefFactory] ?
            (TEntityRefFactory extends Factory<string, any> ?
                (TAction extends string ? [TAction, ReturnType<TEntityRefFactory>] : never)
                : never)
            : never;
}


