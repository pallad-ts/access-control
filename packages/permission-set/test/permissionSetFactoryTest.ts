import {createFactory} from "@pallad/entity-ref";
import {permissionSetFactory} from "@src/permissionSetFactory";
import {IsExact, assert} from "conditional-type-checks";
import {PermissionSet} from "@src/PermissionSet";
import {ORGANIZATION_REF, organizationFactory, PRODUCT_REF, productFactory} from "./fixtures";

describe('permissionSetFactory', () => {
	const factory = permissionSetFactory('product', {
		canUpdate: ['update', productFactory] as const,
		canDelete: ['delete', productFactory] as const,
		canCreate: ['create', organizationFactory] as const
	});

	it('provided methods are assigned to final object and those methods return access control query params with action and subject', () => {
		expect(factory.canDelete(PRODUCT_REF))
			.toMatchSnapshot();

		expect(factory.canCreate(ORGANIZATION_REF))
			.toMatchSnapshot();

		expect(factory.canUpdate(PRODUCT_REF))
			.toMatchSnapshot();
	});

	it('types', () => {
		type ProductEntityRef = ReturnType<typeof productFactory>;
		type OrganizationEntityRef = ReturnType<typeof organizationFactory>;

		type Factory = typeof factory;

		assert<IsExact<Parameters<Factory['canCreate']>[0], OrganizationEntityRef>>(true);
		assert<IsExact<Parameters<Factory['canDelete']>[0], ProductEntityRef>>(true);
		assert<IsExact<Parameters<Factory['canUpdate']>[0], ProductEntityRef>>(true);

		assert<IsExact<Factory['permissionSet'], PermissionSet<
			['update', ProductEntityRef] | ['delete', ProductEntityRef] | ['create', OrganizationEntityRef]
		>>>(true);
	});
});
