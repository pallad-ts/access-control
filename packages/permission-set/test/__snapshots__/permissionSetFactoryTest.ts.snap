// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permissionSetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 1`] = `
{
  "action": "delete",
  "subject": {
    "entityRef": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
    "subjectType": "product",
  },
}
`;

exports[`permissionSetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 2`] = `
{
  "action": "create",
  "subject": {
    "entityRef": EntityRef {
      "data": {
        "id": "organizationId",
      },
      "type": "organization",
    },
    "subjectType": "product",
  },
}
`;

exports[`permissionSetFactory provided methods are assigned to final object and those methods return access control query params with action and subject 3`] = `
{
  "action": "update",
  "subject": {
    "entityRef": EntityRef {
      "data": {
        "id": "productId",
        "organizationId": "organizationId",
      },
      "type": "product",
    },
    "subjectType": "product",
  },
}
`;
